packages/edit-navigation/src/components/sidebar/manage-locations.js

```diff
@@ -31,6 +34,42 @@ export default function ManageLocations( {
 	const [ isModalOpen, setIsModalOpen ] = useState( false );
 	const openModal = () => setIsModalOpen( true );
 	const closeModal = () => setIsModalOpen( false );
+	const { createSuccessNotice, createErrorNotice } = useDispatch(
+		noticesStore
+	);
+
+	const updateMenuLocations = async () => {
+		const batch = createBatch();
+		menus.forEach( ( { id } ) => {
+			const locations = menuLocations
+				.filter( ( menuLocation ) => menuLocation.menu === id )
+				.map( ( menuLocation ) => menuLocation.name );
+
+			batch.add( {
+				path: `/__experimental/menus/${ id }`,
+				data: {
+					locations,
+				},
+				method: 'POST',
+			} );
+		} );
+
+		const isSuccess = await batch.run();
+		if ( isSuccess ) {
+			createSuccessNotice( __( 'Menu locations have been updated.' ), {
+				type: 'snackbar',
+			} );
+			closeModal();
+			return;
+		}
```

Comment:
@anton-vlasenko  The bad news is that `createBatch` is an experimental API and we don't want to export it from core-data – this would make it public API that developers may start relying in.

The good news is that you don't need to do that. You could something along these lines to actions.js:

```js
export const updateMenuLocations = ( menus, menuLocations ) => ({ registry }) => {
	const tasks = menus.map( ( { id } ) => {
		const locations = menuLocations
			.filter( ( menuLocation ) => menuLocation.menu === id )
			.map( ( menuLocation ) => menuLocation.name );

		return ({ saveEntityRecord }) => saveEntityRecord( 'root', 'menu', {
			locations
		} );
	} );

	const results = await registry
		.dispatch( 'core' )
		.__experimentalBatch( tasks );
}
```

And then do something like this to your component:

```js
const { updateMenuLocations } = useDispatch( navigationStore );
const handleUpdate = async () => {
	try {
		await updateMenuLocations( menus, menuLocations );

		createSuccessNotice( __( 'Menu locations have been updated.' ), {
			type: 'snackbar',
		} );
		closeModal();
	} catch( e ) {
		createErrorNotice( __( 'There was an error.' ), {
			type: 'snackbar',
		} );
	}
}
```

You will also need to add [`__experimentalUseThunks: true`](https://github.com/WordPress/gutenberg/pull/34541/files#diff-dd8dc10b4147447239a00e1904b08cfa0dc8d3480ec8fed0bbcdbbd48c13cf5cR30) to store definition.


There are a few caveats here. For once, handling errors is pretty tricky. You will need to check for `getLastEntitySaveError` for each saved menu:

```js
	registry
		.select( 'core' )
		.getLastEntitySaveError( kind, entityType, change.id )
```

Ideally the batch would just throw an exception when things fail, but unfortunately this isn't the case now. We may [rework batching soon](https://github.com/WordPress/gutenberg/pull/34295), but it is what it is for now.

Reply:
We don't have much documentation on __experimentalBatch, but we use batch saving in the edit-widgets package: https://github.com/WordPress/gutenberg/blob/trunk/packages/edit-widgets/src/store/actions.js#L95

Reply:
@adamziel how come it does not have an experimental name?

Reply:
@draganescu I think __experimental is for exported things to warn others.`createBatch()` is not exported anywhere. CC @noisysocks who implemented that

Reply:
Thank you for such a great comment, @adamziel ! I owe you. I've learned a lot from it.
I have a question.
1. Is that refactoring something that is **nice** to have? Or we **must** do it? I'm asking because I'm not sure how much time I'm allowed to spend on this task. I'm not so familiar core-data package.
2. We've [agreed](https://github.com/WordPress/gutenberg/issues/33815#issuecomment-908710847) that we will use `experimental/menus/${menuId}` endpoint to send requests to update menus' locations. Do you know if your suggestion uses that endpoint to update menus under the hood?

Reply:
@anton-vlasenko
1. we **can't** export createBatch() from core-data so this one is a blocker
2. I haven't actually tested that but I think so, refer to entities.js in core-data whenever in doubt. I'll only be able to confirm on Tuesday so might be faster to check

Reply:
Thank you for such a great comment, @adamziel ! I owe you. I've learned a lot from it.
I've refactored my code and now it doesn't depend on the `createBatch` function.
I have a question.
 
1. Is your suggestion to add a new method to the `actions.js` something that is **nice** to have? Or we **must** do it? I'm asking because I'm not sure how much time I'm allowed to spend on this task. I'm not so familiar with the core-data package.
2. We've [agreed](https://github.com/WordPress/gutenberg/issues/33815#issuecomment-908710847) that we will use `experimental/menus/${menuId}` endpoint to send requests to update menus' locations. Do you know if your suggestion uses that endpoint to update menus under the hood?

Reply:
@adamziel I would leave it as it is because I've refactored my PR and it doesn't depend on `createBatch()` now.
If you insist that we need to add a new method to the `actions.js` file and use core-data package, so be it. Please let me know and I'll go that route.
I'd like to leave this PR as it is mainly because of the amount of time I've already spent on it. Time is my main concern here.
Thank again for all the feedback you gave.